### DISCUSSION TITLE
Motion Commander Over Bluetooth Can't Fly Again

### DISCUSSION THREAD

**Discussion by dbaldwin** [bitcraze/discussions#996] [Q&A]
I have been working to mimic the Crazyflie Python library in TypeScript. Mainly so that our users can quickly program their drone from a web browser using JavaScript. I've been able to successfully implement the MotionCommander and the results have been very impressive using Crazyflie with the Flow Deck. We can connect to Crazyflie using Bluetooth from any browser that supports the Bluetooth API. It performs identically to the Python library, expect for one thing. After we run a simple mission such as takeoff, fly forward, and land, we cannot takeoff again. 

Previoiusly @krichardsson submitted a PR for a Python fix where we experienced the same issue:

https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/positioning/motion_commander.py#L122

and we are sending that same packet to Crazyflie at the end of the mission after it lands. But then we can't launch the mission again so I believe something else is at play. We are not currently implementing any of the param/toc packets and I'm wondering if that could be the problem? I noticed that in the python motion commander there is a reset_position_estimator call inside of the takeoff method:

https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/positioning/motion_commander.py#L93

I'm trying to avoid having to rewrite every single class, but am wondering what needs to happen for Crazyflie to be able to take off after a successful mission and landing. At some point we will implement most of the Python lib capabilities, but right now we're focused on testing autonomous flight commands. 

Any thoughts/ideas would be much appreciated.

**Comment by krichardsson** [comment]
The issue that I (hopefully) fixed in the motion commander addressed the problem that the supervisor locks the CF as it thinks that the ground station is lost. We recently added a function to the client that will make it easier to understand if this is the case. Get the latest python client (from master in git) and connect to the CF when you have ended up in the problematic state. Check out the new "supervisor info" field at the bottom left in the flight tab, it will tell you if the CF is locked or not. 

The other possible issue that you might have run into is that it is not possible to mix and match bluetooth and the Crazyradio freely. As soon as the CF receives a packet from the Crazyradio it will disable bluetooth (for performance reasons) and it must be restarted to enable BLE again. This will for instance apply if you do the test above.

**Reply by dbaldwin** [reply]
![crazyflie_client_supervisor](https://github.com/bitcraze/discussions/assets/38339/0f7ac9c9-48a1-4aa5-a9bd-1bb84d80afed)

From what I can tell the supervisor is locked after running a simple mission with our TypeScript code. We are not mixing Bluetooth and Crazyradio. So is it your understanding that if we mimic the send_notify_setpoint_stop packet here:

https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/commander.py#L92

Then we should be able to "unlock" Crazyflie so that we can send more commands again? We will certainly get to work on that, but I want to make sure there aren't any other potential issues at play. Thanks for the assistance.

**Reply by krichardsson** [reply]
Ah, sorry my bad, you also need the latest firmware. The supervisor information is missing in your image from your client, it should look like this when happy
![image](https://github.com/bitcraze/discussions/assets/1464817/07143d8e-c980-4a1a-a384-08c64d3fa79f)

and like this when locked
![image](https://github.com/bitcraze/discussions/assets/1464817/65c1e419-aa78-4bad-9777-f5776600ab1a)


> Then we should be able to "unlock" Crazyflie so that we can send more commands again?

Currently there is no functionality to unlock a crashed Crazyflie, the idea is that you probably want to check the hardware after a crash.

If your platform actually is locked, the question is why? It should not happen normally by taking off, flying and landing. A locked CF should be an exceptional condition that occurs when the CF is upside down, lost connection with the ground etc. 

Maybe you also want to display the supervisor state in your UI in one way or the other, to let the user know that the CF must be restarted?
The best way is to use [bit 0x40 in the supervisor info bit field](https://github.com/bitcraze/crazyflie-firmware/blob/dbb9df1137f11d4e7e3771c56d25a7137b5b69cc/src/modules/src/supervisor.c#L312-L314), you can get it using the [`supervisor.info` log](https://github.com/bitcraze/crazyflie-firmware/blob/dbb9df1137f11d4e7e3771c56d25a7137b5b69cc/src/modules/src/supervisor.c#L452)

**Reply by dbaldwin** [reply]
@krichardsson thanks for the heads up. I did update to the latest firmware and have confirmed that after issuing my commands over Bluetooth does leave CF in a locked state. If you have some time can you give the following a try please? 

https://droneblocks-crazyflie.web.app/

Make sure you're in a Chrome browser if possible (or any browser that supports HTML5 bluetooth). Open up the dev console so you can see the packets I'm sending to Crazyflie. Click connect and in the console you'll see something like "Bluetooth ready to send". Click takeoff and you'll see CF takeoff as well as the setpoint packets being sent at 5hz. You can navigate around and finally click land. This will do the same thing as the Python motion commander... reducing the z_distance and finally sending the stop setpoint as well as the notify setpoint stop packet. Those will show up in the console as well.

Then if you click takeoff again you'll see that CF does nothing. But you will notice that the receive LED on the front left arm is blinking for each message received.

I then took your advice and connected to cfclient via CrazyRadio and yes the supervisor reports as locked.

Any insight or guidance would be very much appreciated. After we get over this hurdle we are going to be building the DroneBlocks interface to CrazyFlie as well as introduce gamepad support. Most of our classroom users are on ChromeBooks and won't be able to install drivers for CrazyRadio so Bluetooth is out best bet.

I will continue to investigate from this end. Thanks for your support!

**Reply by knmcguire** [reply]
Just double checking before kristoffer looks at this, did you manage to implement the  send_notify_setpoint_stop packet or haven't you tried it yet?

**Reply by dbaldwin** [reply]
Yes, sending that packet. See the screenshot below that shows a recent flight where I issued a takeoff and land sequence. The land method mimics what's in the motion commander python class. It ends with the notify stop setpoint packet...

![crazyflie_setpoint](https://github.com/bitcraze/discussions/assets/38339/4f57e76b-0cfa-4d86-af44-8a8d01e012d4)

Feel free to go to this link:

https://droneblocks-crazyflie.web.app/

open your console and click takeoff and then land. You don't need to connect to Crazyflie to see the packets being logged. I'm certain we're missing something here (on our end) unless you think it may be something unique to Bluetooth communication and the motion commander?

**Reply by krichardsson** [reply]
I looked a bit at this and it seems as the problem is that the `sendNotifySetpointStop()` packet never arrives in the CF, but I don't know why. The `sendStopSetpoint()` is received as expected but nothing after this packet. 
Your js code looks reasonable so I'm not sure what is going on. 

The communication link is originally designed to have a constant flow of packets, the two way communication is driven from the client. This is implemented through the client [sending a null packet](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/#null-packet-and-special-packets) to poll data from the Crazyflie and empty out queues. I don't think this is what is causing the problems you see, but I believe this is something you should add. Sending the null packet 50 - 100 times/s is probably enough (you might have to do some experimenting), it would enable you to use the logging framework if you want to, and it would keep the LED blinking on the CF to notify the user that the link is active.

**Reply by dbaldwin** [reply]
I've done a bit of hack workaround for the time being. Basically I was working on gamepad support and looking into how to blend the two. Where a user could be programming, switch to manual flight with a gamepad, and then switch back to programming. In doing that I learned of the standard `send_setpoint` commander method. So what I've opted to do for the time being is once CF lands I start sending a zero setpoint at 5hz. That keeps CF alive so now I can takeoff again and control using hover setpoints. Feel free to give it a try here:

https://droneblocks-crazyflie.web.app/

Also, if you happen to have a FlySky i6s transmitter you can fly CF manually by plugging in your TX and going here:

https://droneblocks-crazyflie.web.app/gamepad

I will be working on adding other gamepads and this was just a proof of concept with some hardware laying around.

Unless you think there's any risk to my approach above I say we close this discussion out. Or if you think I should by sending the null packet instead of the zero setpoint then I could give that a try. I didn't see any examples of sending a null packet in the python lib and a lot of what I'm doing is mirroring what is done there in my Typescript lib.

**Reply by krichardsson** [reply]
Your solution is fine, this is pretty much what is happening in the python client, it simply sends setpoints with no thrust but never the `sendNotifySetpointStop()` packet. `sendNotifySetpointStop()` re-enables the high level commander internal "sending" of setpoints but an equally good option is that the client does it. It is also enough to send setpoints for a couple of seconds after landing (actually until the supervisor understands that the CF has landed), after that it is not important until the next flight. 

The sending of null packets [is done in the radio driver](https://github.com/bitcraze/crazyflie-lib-python/blob/c1c42f7a8b58bc95d660247d5a2a6832755dae90/cflib/crtp/radiodriver.py#L670) (I think) and is not very obvious.
